File: SymmetricTree.py
# Definition for a  binary tree node
class Node:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    # @param root, a tree node
    # @return a boolean
    def isSymmetric(self, root):
        if root == None:
            return True
        return self.isSym(root.left, root.right)
            
    def isSym(self, r1, r2):
        if not r1 and not r2:
            return True
        if not r1 or not r2:
            return False
        if r1.val != r2.val:
            return False
        return self.isSym(r1.left, r2.right) and self.isSym(r1.right, r2.left)

    def isSymmetric2(self, root):
        if root == None:
            return True
        stk1, stk2 = [], []
        p, q = root.left, root.right
        # not ( p && q )
        if not p and not q:
            return True
        if (p and not q) or (not p and q): 
            return False
        stk1.append(p), stk2.append(q)
        while stk1 and stk2:
            p, q = stk1.pop(), stk2.pop()
            if p.val != q.val:
                return False
            if (p.right and not q.left) or (not p.right and q.left):
                return False
            if p.right and q.left:
                stk1.append(p.right)
                stk2.append(q.left)
                
            if (p.left and not q.right) or (not p.left and q.right):
                return False
            if p.left and q.right:
                stk1.append(p.left)
                stk2.append(q.right)
        return True
            

        
if __name__ == '__main__':
    n1 = Node(2, Node(3), Node(4))
    n2 = Node(2, Node(4), Node(3))
    #root = Node(1, n1, n2)
    root = Node(1)
    so = Solution()
    print so.isSymmetric(root)
    print so.isSymmetric2(root)

File: CPP/Second/symmetric_tree.cc
// https://oj.leetcode.com/problems/symmetric-tree/

// Given a binary tree, check whether it is a mirror of itself (ie, symmetric
// For example, this binary tree is symmetric:
//     1
//    / \
//   2   2
//  / \ / \
// 3  4 4  3
// But the following is not:
//     1
//    / \
//   2   2
//    \   \
//    3    3
// Note:
// Bonus points if you could solve it both recursively and iteratively.
// confused what "{1,#,2,3}" means? > read more on how binary tree is serialized
// OJ's Binary Tree Serialization:
// The serialization of a binary tree follows a level order traversal, where '#'
// Here's an example:
//    1
//   / \
//  2   3
//     /
//    4
//     \
//      5
// The above binary tree is serialized as "{1,2,3,#,#,4,#,#,5}".

/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;

class Solution {
public:
    bool isSym(TreeNode *p, TreeNode * q) {
        if (!p && !q)
            return true;
        if (!p || !q)
            return false;
        if (p->val != q->val)
            return false;
        return isSym(p->left, q->right) && isSym(p->right, q->left);
    }
    
    bool isSymmetric(TreeNode *root) {
        if (!root)
            return true;
        return isSym(root->left, root->right);
    }
   
};

int main(int argc, char *argv[])
{

    return 0;
}
